### ISSUE TITLE
Out of Memory Error [GC overhead, java heap space] occurs very often

### ISSUE THREAD

**Issue by gdsentropy** [processing/processing#4510]
Hello,

After updating to v3.1.1 on Mac OS X ElCapitan, I am experiencing vast numbers of out of memory errors, like the following:

Exception in thread "AWT-EventQueue-0" java.lang.OutOfMemoryError: GC overhead limit exceeded
Exception in thread "AWT-EventQueue-0" java.lang.OutOfMemoryError: Java heap space

It got so bad I had to go back to using 3.0.2, which does not have these issues.

Ian

**Comment by JakubValtar** [comment]
Hi @gdsentropy, thanks for the report and sorry for the trouble. Lot of backend code changed in 3.1, however from what I've seen the memory consumption mostly improved.

I wonder what could be special about your system. I don't know why this would happen, but I have a few questions which may help me identify the problem better.

Could you describe when the OutOfMemoryError happens?
How many Processing windows are open?
How big is your sketch (number of tabs and lines)?
How many libraries you have installed and how many are used in your sketch?
Do you have any libraries which are not from Contribution Manager (e.g. installed manually)?
Did you put anything into the "code" folder of the sketch? (sketchFolder/code)

Sorry about the rectangular selection, but we decided to remove it, because (as you noticed) it was implemented only half-way and despite some efforts we are not able to fix it and maintain it.

**Comment by gdsentropy** [comment]
>>Could you describe when the OutOfMemoryError happens?

Sure.

>>How many Processing windows are open?

It happens the same frequency if I have one or ten, though I am more likely to have ten than one. Copy/paste will break when moving or duplicating a selection from one window to another quite often, and this will happen with even the minimum of the two windows needed. This produces some thread error after locking all windows for a few minutes; I do not recall the precise issue but it is in my .err files. Sometimes one of the other two errors follows. This happens on both mac and windows.

>>How big is your sketch (number of tabs and lines)?

My sketches typically have between 4 and 20 tabs, and many hundreds of lines of code to a few thousand. I code reaction-diffusion simulations.

>>How many libraries you have installed and how many are used in your sketch?

Zero and zero; I write all of my own code, and generally do not even make use of any parts of the Java framework because I like to be able to move my code between C#, Java and C++ at will with minimum modifications.

>>Do you have any libraries which are not from Contribution Manager (e.g. installed manually)?

None.

>>Did you put anything into the "code" folder of the sketch? (sketchFolder/code)

No; all code executed within my sketches is written in Processing. All classes are written in Processing and end up as tabs.

I do use classes in every sketch, even though it seems that in Processing they end up as classes nested into whatever Core is.

I am starting to think that I really do need to just abandon the editor. 

I ported over the sections of Core that I use (and I do not use much of it, mostly just the unary math operators) to C# for when I migrate the simulations I prototype in Processing to my .NET simulation framework. I do the same for my Java framework as well.

The editor is useful to me in the capacity of being able to quickly prototype the simulations I write which produce my art, and depending on if I am using a forward euler based PDE solver or a non-PDE solver which simulates the behavior of a PDE, I have small sketch frameworks I wrote to support each.

Thanks,
Ian

**Comment by JakubValtar** [comment]
Thanks for this info!

Hm, this is strange, recently I was testing a sketch with 15,000 lines of code and ten libraries and although the error checking was little bit slower the editor was still perfectly usable.

Could you share the error files with me?

It would be best if you could send me some sketch which crashes the editor (I would of course keep it confidential). If you want to contact me directly my email is on my profile.

I guess that's all I can do right now.

**Comment by StanLepunK** [comment]
@gdsentropy  It's reason whey I use Sublim Text with Processing, for this problem of memory, That's happen often with the version alpha of Processing 3, now I have the feeling is more stable with less crash. But that's happen some time when you have few sketches open in same time, that you're case ?

I try to reproduce the bug, with my project, when I have one big sketch 5K lines and one huge with 27K lines that's ok, when I open a third huge sketch Processing freeze, or if it work when I try to run a huge sketch Processing return the crash ````Exception in thread "Thread-16" java.lang.OutOfMemoryError: GC overhead limit exceeded ...blablah...````
NB :  After the crash I must restart Processing with icon not from the guilty sketch, because Processing show me all the time the window with "Serious Problem" and close.
![capture d ecran 2016-05-30 a 16 15 03](https://cloud.githubusercontent.com/assets/3332269/15651820/1f4fe19c-2682-11e6-9691-f4017ac7ab11.png)


@JakubValtar I know this code with 15k lines and a lot of libraries imported :) Now is 27K lines of codes

**Comment by gdsentropy** [comment]
@JakubValtar

>>Could you share the error files with me?

Absolutely. The zip file contains two 7Z files; the Mac one contains over 300mb of error output, the Windows one 100 times less, as I primarily use my Windows machines for .NET programming.

>>It would be best if you could send me some sketch which crashes the editor (I would of course keep it confidential). If you want to contact me directly my email is on my profile.

I can send you things, but the behavior is fairly agnostic toward any particular sketch, and as thus I would be flat out guessing as to an appropriate file to send.

I will look through my .err files, to see if it logs the sketch which caused the error [one would think such would be the case]. If the sketch name is logged, I will mine the files for the most frequent offender, and send it.

Please keep in mind that I use Processing to rapidly prototype algorithms for my art, so the code isn't pretty and I would personally fire any developer who checked in anything even remotely this ugly. But then again, the targets of my code analysis consider me a bit of a harass and dread being brought in front of me in meetings....so who knows, maybe it isn't so bad.

@StanLepunK

I get the gift of the error prompt window you attached at least a few times per week as well. Isn't that failure mode just wonderful?

Generally with any of the exceptions mentioned in this thread, the processing instance gets hosed and must be completely nuked before it takes down the entire machine...which it will do.

The ones brought on by trans-window copy/paste are the most annoying, because they prevent even saving of your current work.

As I said, all of this adds up to my pending abandonment of Editor for IntelliJ once I figure out how to build the IntelliJ equivalent of VSIX templates, but that isn't any reason for me to keep these errors to myself, given that it will result in the net improvement of the package for everyone. 

Maybe one day my beloved Rectangular Selections will come back. ;) I would issue a monetary bounty for that feature if I thought it would do any good, and I am not kidding just in case someone is interested...

Ian
[Debug and Console Folders [Mac and Windows].zip](https://github.com/processing/processing/files/289838/Debug.and.Console.Folders.Mac.and.Windows.zip)

**Comment by JakubValtar** [comment]
Thanks, I will take a look.

> Please keep in mind that I use Processing to rapidly prototype algorithms for my art, so the code isn't pretty and I would personally fire any developer who checked in anything even remotely this ugly.

Haha, I do livecoding, I've seen it all.

> As I said, all of this adds up to my pending abandonment of Editor for IntelliJ once I figure out how to build the IntelliJ equivalent of VSIX templates, but that isn't any reason for me to keep these errors to myself, given that it will result in the net improvement of the package for everyone.

Just export your sketch and you will get a .java file which you can add to IntelliJ project and add core.jar as a dependency (and remaining jars from that folder if you use opengl). Or you can export your framework as a jar and drop in into code folder of your sketches. It should make processing editor much happier.

**Comment by gdsentropy** [comment]
I know we are drifting OT now [and I don't really care] though I'll go ahead and reply regardless:

> Just export your sketch and you will get a .java file which you can add to IntelliJ project and add core.jar as a dependency (and remaining jars from that folder if you use opengl). Or you can export your framework as a jar and drop it into code folder of your sketches. It should make processing editor much happier.

Sigh...this is all going in the direction of nullifying the entire point of using Processing to begin with for me, that being that it is/was something that I could fire up in a few seconds and slam an algorithm out with, when in the middle of the night I get woken up or am prevented from sleeping by an explosion of math, code and images. I really love Processing and I don't like this fact.

I am not sure if you are familiar with Visual Studio to know what I meant by VSIX project template, but to put it simply, for IntelliJ I need to find out how to build a template that lets me just create a new project with everything I need already in it in under 10s from the time the IDE loads, so far I have not discovered how to do this, as Google isn't quite as useful for IntelliJ as it is for Visual Studio. :(

Do you know if there is some place to post feature bounties for Processing? If there isn't, there should be. 

I'd throw $100USD at the rectangular selection feature, negotiable. It would cost far more in time to do it myself, plus there is the fact my dev boxes aren't setup to build Processing and it is an unfamiliar codebase...

I will look up your e-mail, if I can find my way around Github. I only joined for Processing, but I may leave Github due to political conflict with their ideologies. Oh well.

Ok, enough OT from me now.

**Comment by softpunch** [comment]
Any progress on this?

I have a similar problem on El Capitan.

Memory errors just happen frequently, 
even if I'm running very small sketches with no libraries.
